save-orchestrator/src/main/kotlin/org/cqfn/save/orchestrator/service/AgentService.kt

```diff
@@ -230,15 +234,62 @@ class AgentService {
     private fun TestBatch.toHeartbeatResponse(agentId: String) =
             if (tests.isNotEmpty()) {
                 // fixme: do we still need suitesToArgs, since we have execFlags in save.toml?
-                NewJobResponse(
-                    tests,
-                    suitesToArgs.values.first() +
-                            " " + tests.joinToString(separator = " ") { it.filePath }
-                )
+                constructCliCommand(tests, suitesToArgs).flatMap { cliArgs ->
+                    Mono.fromCallable {
+                        NewJobResponse(tests, cliArgs)
+                    }
+                }
             } else {
                 log.info("Next test batch for agentId=$agentId is empty, setting it to wait")
-                WaitResponse
+                Mono.fromCallable {
+                    WaitResponse
+                }
+            }
+
+    @Suppress("TOO_LONG_FUNCTION")
+    private fun constructCliCommand(tests: List<TestDto>, suitesToArgs: Map<Long, String>): Mono<String> {
+        var isStandardMode = false
+        val testPaths: MutableList<String> = mutableListOf()
+        // first, need to check the current mode, it could be done by looking of type of any test suite for current tests
+        return webClientBackend.get()
+            .uri("/testSuite/${tests.first().testSuiteId}")
+            .retrieve()
+            .bodyToMono<TestSuite>()
+            .flatMap { testSuite ->
+                isStandardMode = testSuite.type == TestSuiteType.STANDARD
+                if (isStandardMode) {
+                    // in standard mode for each test get proper prefix location, since we created extra directories
+                    Flux.fromIterable(tests).flatMap { test ->
```

Comment:
Group `tests` by `testSuiteId` to avoid a lot of equivalent requests?

Reply:
Make sense, changed